client/ayon_core/plugins/publish/extract_usd_layer_contributions.py

```diff
@@ -0,0 +1,854 @@
+# We need to at least know what it targets (e.g. where does it go into) and
+# in what order (which contribution is stronger?)
+# Preferably the bootstrapped data (e.g. the Shot) preserves metadata about
+# the contributions so that we can design a system where custom contributions
+# outside of the predefined orders are possible to be managed. So that if a
+# particular asset requires an extra contribution level, you can add it
+# directly from the publisher at that particular order. Future publishes will
+# then see the existing contribution and will persist adding it to future
+# bootstraps at that order
+# TODO: Avoid hardcoded ordering - might need to be set through settings?
+LAYER_ORDERS = {
+    # asset layers
+    "model": 100,
+    "assembly": 150,
+    "groom": 175,
+    "look": 200,
+    "rig": 300,
+    # shot layers
+    "layout": 200,
+    "animation": 300,
+    "simulation": 400,
+    "fx": 500,
+    "lighting": 600,
+}
+
+# This global toggle is here mostly for debugging purposes and should usually
+# be True so that new publishes merge and extend on previous contributions.
+# With this enabled a new variant model layer publish would e.g. merge with
+# the model layer's other variants nicely, so you can build up an asset by
+# individual publishes instead of requiring to republish each contribution
+# all the time at the same time
+BUILD_INTO_LAST_VERSIONS = True
+
+
+@dataclasses.dataclass
+class _BaseContribution:
+    # What are we contributing?
+    instance: pyblish.api.Instance  # instance that contributes it
+
+    # Where are we contributing to?
+    layer_id: str  # usually the department or task name
+    target_product: str  # target product the layer should merge to
+
+    order: int
+
+
+class SublayerContribution(_BaseContribution):
+    """Sublayer contribution"""
+
+
+@dataclasses.dataclass
+class VariantContribution(_BaseContribution):
+    """Reference contribution within a Variant Set"""
+
+    # Variant
+    variant_set_name: str
+    variant_name: str
+    variant_is_default: bool  # Whether to author variant selection opinion
+
+
+def get_representation_path_in_publish_context(
+    context: pyblish.api.Context,
+    project_name,
+    folder_path,
+    product_name,
+    version_name,
+    representation_name,
+):
+    """Return resolved path for product if present in publishing context.
+
+    Allow resolving 'latest' paths from a publishing context's instances
+    as if they will exist after publishing without them being integrated yet.
+
+    Args:
+        context (pyblish.api.Context): Publishing context.
+        project_name (str): Project name.
+        folder_path (str): Folder path.
+        product_name (str): Product name.
+        version_name (str): Version name.
+        representation_name (str): Representation name.
+
+    Returns:
+        Union[str, None]: Returns the path if it could be resolved
+
+    """
+    if context.data["projectName"] != project_name:
```

Comment:
Not that important, but why?

Reply:
A USD file may contain references to outside of the current project, e.g. from a library project. We know however, that due to how the publishing logic works that the current publish context never publishes instances into OTHER projects - hence, for those paths... we do not need to check further inside the publishing context since it'd be a redundant check.

Reply:
Added a comment to the code.